src/worker/tasks/chainIndexer.ts

```diff
@@ -285,67 +284,63 @@ export const createChainIndexerTask = async (
           }
 
           const sdk = await getSdk({ chainId });
-
           const provider = sdk.getProvider();
-          const currentBlockNumber = await provider.getBlockNumber();
 
-          // check if up-to-date
-          if (lastIndexedBlock >= currentBlockNumber) {
+          // Get latest block with logs. This should be the maxBlock to query up to.
+          const logs = await provider.getLogs({ fromBlock: "latest" });
+          if (logs.length === 0) {
             return;
           }
+          const currentBlockNumber = logs[0].blockNumber;
 
-          // limit max block numbers
-          let toBlockNumber = currentBlockNumber;
-          if (currentBlockNumber - (lastIndexedBlock + 1) > maxBlocksToIndex) {
-            toBlockNumber = lastIndexedBlock + 1 + maxBlocksToIndex;
+          // Check if up-to-date.
+          if (lastIndexedBlock >= currentBlockNumber) {
+            return;
           }
 
+          // Limit max block number.
+          const toBlockNumber = Math.min(
+            currentBlockNumber,
+            lastIndexedBlock + maxBlocksToIndex,
+          );
+
           const subscribedContracts = await getContractSubscriptionsByChainId(
             chainId,
           );
-          const subscribedContractAddresses = [
-            ...new Set<string>(
-              subscribedContracts.map(
-                (subscribedContract) => subscribedContract.contractAddress,
-              ),
+          const subscribedContractAddresses = dedupeArray(
+            subscribedContracts.map(
+              (subscribedContract) => subscribedContract.contractAddress,
             ),
-          ];
+          );
 
           await Promise.all([
+            // Checks eth_getLogs.
             indexContractEvents({
               pgtx,
               chainId,
-              fromBlockNumber: lastIndexedBlock + 1,
+              fromBlockNumber: lastIndexedBlock,
```

Comment:
Potential fix 3: Overlap the "boundary" blocks just in case there's incomplete data from RPC